gossipsub.go

```diff
@@ -917,6 +967,27 @@ func (gs *GossipSubRouter) handlePrune(p peer.ID, ctl *pb.ControlMessage) {
 	}
 }
 
+func (gs *GossipSubRouter) handleIDontWant(p peer.ID, ctl *pb.ControlMessage) {
+	if gs.unwanted[p] == nil {
+		gs.unwanted[p] = make(map[checksum]int)
+	}
+
+	// IDONTWANT flood protection
```

Comment:
We need global maximums as well, right? Or do we limit the number of grafted peers?

Reply:
that is limited by the mesh size, so we probably don't need a global maximum.

Reply:
however, we should probably ignore IDONTWANT's from peers that are not in the mesh -- so there is something to fix here.

Reply:
Alright, let's ignore IDONTWANT's from non-mesh peers.

Reply:
Wait. IDONTWANT is not topic-specific and there is no way to discover the topic by looking at the IDONTWANT messages, so I think it's better not to handle this.

Reply:
We don't have to make it topic specific, just only accept said messages from peers to which we are grafted (on any topic) and, ideally, limit the number of IDONTWANTs per peer.

Reply:
there is and if it is exceeded it gets cleared on next heartbeat.

Reply:
But I don't see any limits on the number of peers.

Also, we're currently dealing with an issue in F3 where we're getting absolutely _hammered_ by IHAVE messages on topics we're not subscribed to.

Reply:
yeah, maybe we should track the topic in the cache and avoid sending IHAVEs if the peer is not in the topic.

But thats a larger change, that will probably need a spec update, and certainly out of scope of this pr.

Do you want to open an issue about this, maybe in specs?

Reply:
as for the peers, the meshes are kept within degree bounds, maintained in the heartbeat.

So i am not convinced it is an effective  attack vector.

Lets open a follow up issue for this?

Reply:
> Also, we're currently dealing with an issue in F3 where we're getting absolutely _hammered_ by ~IHAVE~ IWANT messages on topics we're not subscribed to.

Could reference the issue on IWANT? or explain it in detail. I can't imagine how you can be dos'ed by receiving too many IWANTs (or IHAVEs).

Reply:
> yeah, maybe we should track the topic in the cache and avoid sending IHAVEs if the peer is not in the topic.

I will be surprised if go-libp2p sends IHAVEs to some peers that are not in the topic. I think it's probably already in the spec and I always thought that no implementation sends IHAVEs to non-subscribing peers.

Reply:
> I will be surprised if go-libp2p sends IHAVEs to some peers that are not in the topic. I think it's probably already in the spec and I always thought that no implementation sends IHAVEs to non-subscribing peers.

Doesn't the spec already contemplate sending messages on topics that it isn't subscribed to using the fanout peers?

I know that this is a common practice in Ethereum CL, and I don't see any kind of filter on the [GossipSub publishing side](https://github.com/libp2p/go-libp2p-pubsub/blob/19ffbb3a482caecabb8520917c631e3047a78094/gossipsub.go#L1037) that prevents adding a message to the cache of messages if we are not subscribed to the topic 🤷🏽

Reply:
I think it might make sense to specify the behavior.
It makes more sense to track the topic and only gossip IHAVEs to peers in the topic.

Reply:
At any rate, it is out of scope of this pr.

Reply:
> At any rate, it is out of scope of this pr.

yeah, let's close this thread

Reply:
@cortze so it's like sending IHAVEs after publishing (not forwarding) a message using the fanout? got it thanks, but my intuition is that the fanout publishers shouldn't consequently send IHAVEs. Having only the forwarders send them should be enough.

Reply:
That should be something that could be discussed. If you published the message, even if you are not subscribed to the topic, you definitely have it and I don't see why you wouldn't share it on an IHAVE (assuming that the remote peer IS subscribed to the topic).
I was just introducing the edgy case where you could receive IWANT messages on topics that you are not subscribed to.